scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+      ): Task[Unit] = {
+        val caller = (remotePublicKey, channel.to)
+
+        payload match {
+          case Ping(_, _, to, _, maybeRemoteEnrSeq) =>
+            maybeRespond {
+              handler.ping(caller)(maybeRemoteEnrSeq)
+            } { maybeLocalEnrSeq =>
+              channel.send(Pong(to, hash, 0, maybeLocalEnrSeq)).void
+            }
+
+          case FindNode(target, expiration) =>
+            maybeRespond {
+              handler.findNode(caller)(target)
+            } { nodes =>
+              nodes
+                .take(kademliaBucketSize)
+                .grouped(maxNeighborsPerPacket)
+                .toList
+                .traverse { group =>
+                  channel.send(Neighbors(group.toList, 0))
+                }
+                .void
+            }
+
+          case ENRRequest(_) =>
+            maybeRespond {
+              handler.enrRequest(caller)(())
+            } { enr =>
+              channel.send(ENRResponse(hash, enr)).void
+            }
+        }
+      }
+
+      private def maybeRespond[Res](maybeResponse: Task[Option[Res]])(
+          f: Res => Task[Unit]
+      ): Task[Unit] =
+        maybeResponse.attempt.flatMap {
+          case Right(Some(response)) =>
+            f(response)
+          case Right(None) =>
+            Task.unit
+          case Left(NonFatal(ex)) =>
+            // Not responding to this one, but it shouldn't stop handling further requests.
+            Task(logger.error(s"Error handling incoming request: $ex"))
+        }
+
+      private def pack(payload: Payload): Task[Packet] =
+        Packet
+          .pack(payload, privateKey)
+          .fold(
+            err => Task.raiseError(new IllegalArgumentException(s"Could not pack $payload: $err")),
+            packet => Task.pure(packet)
+          )
+
+      private def setExpiration(payload: Payload): Task[Payload] = {
+        payload match {
+          case p: Payload.HasExpiration[_] =>
+            currentTimeMillis.map(t => p.withExpiration(t + expirationMillis))
+          case p =>
+            Task.pure(p)
+        }
+      }
+
+      override val ping = (to: A) =>
+        (localEnrSeq: Option[ENRSeq]) =>
+          peerGroup.client(to).use { channel =>
+            channel
+              .send(
+                Ping(version = 4, from = localNodeAddress, to = toNodeAddress(to), 0, localEnrSeq)
+              )
+              .flatMap { packet =>
+                channel.collectFirstResponse {
+                  case Pong(_, pingHash, _, maybeRemoteEnrSeq) if pingHash == packet.hash =>
+                    maybeRemoteEnrSeq
+                }
+              }
+          }
+
+      override val findNode = (to: A) =>
+        (target: PublicKey) =>
+          peerGroup.client(to).use { channel =>
+            channel.send(FindNode(target, 0)).flatMap { _ =>
+              channel.collectAndFoldResponses(kademliaTimeout, Vector.empty[Node]) {
+                case Neighbors(nodes, _) => nodes
```

Comment:
from what i see geth sanitizes received nodes based on the fact if they are received from local node or remote node: 
https://github.com/ethereum/go-ethereum/blob/723bd8c17fdf1614d81503efdca9e7eece3766a6/p2p/discover/v4_udp.go#L727.
Maybe we can create the task to do the same ?

Reply:
Yes, thanks, I'll make a ticket.